### PR TITLE
Add Catch test framework and SCSI communication unit test

### PR DESCRIPTION
**Pull request by jmwilson**
This incorporates the Catch testing framework (https://github.com/catchorg/Catch2) and integrates it with the autotools setup. Catch is under the BSL, which is compatible with GPL for stenc's code. New tests can be added by adding them in `tests/Makefile.am` and the tests can be run with `make check`.

The added unit test (`tests/scsi.cpp`) adds coverage for several SCSI commands used by stenc. Changes to other code have been kept to a minimum so the tests reflect the behavior prior to future cleanup and changes: refactoring `SCSIWriteEncryptOptions` to be able to get the parameters buffer for comparison, fixing formatting with debug output as discovered while extracting the SCSI communication, and adding `const` modifiers as is helpful.

### CODE REVIEW

**Comment by jonasstein** [comment]
see also https://github.com/scsitape/stenc/issues/6

File: src/scsiencrypt.cpp [hunk 1]
```diff
@@ -369,17 +373,19 @@ bool SCSIExecute(std::string tapedrive, unsigned char *cmd_p, int cmd_len,
...
 #ifdef DEBUGSCSI
-  std::cout << "SCSI Command: ";
+  std::cerr << "SCSI Command: ";
   for (int i = 0; i < cmd_len; i++) {
-    std::cout << std::hex << cmd_p[i];
+    std::cerr << std::hex << std::setw(2) << std::setfill('0')
+              << (unsigned int) {cmd_p[i]} << ' ';
   }
-  std::cout << "\n";
+  std::cerr << std::endl;
```
Comment (jonasstein) [inline src/scsiencrypt.cpp:381]: Why do you use `endl` here? Afaik `std::cerr` is unbuffered and flushed with each and every output operation.

File: src/scsiencrypt.cpp [hunk 2]
```diff
@@ -369,17 +373,19 @@ bool SCSIExecute(std::string tapedrive, unsigned char *cmd_p, int cmd_len,
 #error "OS type is not set"
 #endif
 #ifdef DEBUGSCSI
-  std::cout << "SCSI Command: ";
+  std::cerr << "SCSI Command: ";
```
Comment (jonasstein) [inline src/scsiencrypt.cpp:376]: Why should this go to `cerr` instead of cout?

File: src/scsiencrypt.cpp [hunk 3]
```diff
@@ -194,66 +194,61 @@ SCSI_PAGE_INQ *SCSIGetInquiry(std::string tapeDevice) {
...
   SSP_PAGE_SDE options;
   // copy the template over the options
   memset(&options, 0, sizeof(SSP_PAGE_SDE));
   byteswap((unsigned char *)&options.pageCode, 2, 0x10);
-  int pagelen = sizeof(SSP_PAGE_SDE);
+  size_t pagelen = sizeof(SSP_PAGE_SDE);
   options.scope = 2; // all IT nexus = 10b
-  options.RDMC = eOptions->rdmc;
-  options.ckod = (eOptions->CKOD) ? 1 : 0;
+  options.RDMC = eOptions.rdmc;
```
Comment (jonasstein) [inline src/scsiencrypt.cpp:205]: these rewrites should be part of a separate PR. Ideally one PR for adding tests only and one PR for each change.
Else we can introduce new mistakes quickly.

File: src/scsiencrypt.cpp [hunk 4]
```diff
@@ -369,17 +373,19 @@ bool SCSIExecute(std::string tapedrive, unsigned char *cmd_p, int cmd_len,
...
 
-  std::cout << "SCSI Data: ";
+  std::cerr << "SCSI Data: ";
   for (int i = 0; i < dxfer_len; i++) {
-    std::cout << std::hex << (dxfer_p[i]);
+    std::cerr << std::hex << std::setw(2) << std::setfill('0')
+              << (unsigned int) {dxfer_p[i]} << ' ';
   }
-  std::cout << std::endl;
+  std::cerr << std::endl;
```
Comment (jonasstein) [inline src/scsiencrypt.cpp:388]: no flush needed here

**Review by jonasstein — CHANGES_REQUESTED** [review]
I think we can merge this with minor changes the `->` to `.` rewrite would better match in a separate PR to keep a better overview, but for me it would be ok this time. @ninthclowd  what do you think?

**Comment by jmwilson** [comment]
Will have a look at the problems with make distcheck and update.

File: src/scsiencrypt.cpp (hunk 1 above)
Comment (jmwilson) [reply src/scsiencrypt.cpp:381]: Fair point; this mostly personal preference and made to match the other case.

File: src/scsiencrypt.cpp (hunk 2 above)
Comment (jmwilson) [reply src/scsiencrypt.cpp:376]: It's diagnostic and should go to stderr. I'd actually say most of the print statements in this program are status messages that are not "output" and should go to stderr.

**Comment by jonasstein** [comment]
You could make a **minimal PR** just adding catch and probably one small test. 
Then I can quickly merge that one and we can continue step by step.

**Comment by jonasstein** [comment]
@jmwilson If you want, we can have a short chat on  https://jitsi.fem.tu-ilmenau.de/stenc now, to coordinate

File: src/scsiencrypt.cpp (hunk 2 above)
Comment (ninthclowd) [reply src/scsiencrypt.cpp:376]: Some users have written script that rely on the output going to stdout.  This would be breaking changes for them.  I think I would prefer that any breaking changes go in a separate PR all together.

File: src/scsiencrypt.cpp (hunk 3 above)
Comment (ninthclowd) [reply src/scsiencrypt.cpp:205]: I tend agree with @jonasstein.  If we could split the tests from the functionality changes that would be awesome.  Thanks for contributing this @jmwilson!

File: src/scsiencrypt.cpp (hunk 2 above)
Comment (ninthclowd) [reply src/scsiencrypt.cpp:376]: To clarify, I think these changes are the right move, but I would prefer them in a new minor (or major) version

**Comment by jmwilson** [comment]
Stripped down to just the minimal changes for adding working tests, and added some comments on the purpose of the individual tests. Left in a few new const modifiers because C++ does not like casting away const that would happen in the unit tests.
